kernel/src/scan/mod.rs

```diff
@@ -195,24 +214,20 @@ impl Scan {
         Ok(scan_action_iter(
             engine,
             log_iter,
-            &self.read_schema,
+            &self.logical_schema,
             &self.predicate,
         ))
     }
 
     /// Get global state that is valid for the entire scan. This is somewhat expensive so should
     /// only be called once per scan.
     pub fn global_scan_state(&self) -> GlobalScanState {
-        let partition_columns = &self.snapshot.metadata().partition_columns;
-        let (_all_fields, read_fields, _have_partition_cols) =
-            get_state_info(self.schema().as_ref(), partition_columns);
-        let read_schema = StructType::new(read_fields);
-        let logical_schema = self.schema().as_ref().clone();
         GlobalScanState {
             table_root: self.snapshot.table_root.to_string(),
-            partition_columns: partition_columns.clone(),
-            logical_schema,
-            read_schema,
+            partition_columns: self.snapshot.metadata().partition_columns.clone(),
+            logical_schema: self.logical_schema.as_ref().clone(),
+            read_schema: self.physical_schema.as_ref().clone(),
```

Comment:
aside: Is there a particular reason `GlobalScanState` takes `Schema` instead of `SchemaRef`? 
Why force a clone like this?

Reply:
Yeah, the idea is that `GlobalScanState` should be (de)serializable so it would be easy for multi-node systems to send between nodes. `Schema` is `Serializable` and `SchemaRef` isn't so it makes it be less code.

We could add a custom `De/Serialize` for it and just go into the arc and serialize the inner, but it wasn't clear to me that complexity was worth it (yet...)

Reply:
Serialization is an engine design problem, not a kernel problem. At most kernel might provide a default serializer mechanism -- but even then I don't know what we'd provide, given the wide variety of ways engines handle serialization? And anyway this is useless for FFI because we don't even expose our own schema type to extern engines.

Meanwhile, large schemas (10k+ columns) could be pretty expensive to clone?

Reply:
We would need to provide a `de/serialize_global_state` ffi call yes. If we want to let engines handle serialization all by themselves, we'll need to have more complex APIs to allow reconstruction of state that the kernel understands. possible but something we'll need to design.

ideally we'll sort most of this out with the "single expression" mode of fixup.

I've added #216 to at least get the `clones` gone, and then we can iterate further.